SG crashes upon providing invalid bucket username/password

### Sync Gateway version
2.0.0-687

### Operating system
CentOS 7

### Config file
```
{
    "interface":":4984",
    "adminInterface": "0.0.0.0:4985",
    "maxIncomingConnections": 0,
    "maxCouchbaseConnections": 16,
    "maxFileDescriptors": 90000,
    "slowServerCallWarningThreshold": 500,
    "compressResponses": false,
    "log": ["CRUD+", "Cache+", "HTTP+", "Changes+", "Import+"],
    "databases":{
        "db":{
            "unsupported" : {
                "replicator_2" :true
	    },
            "import_docs": "continuous",
            "enable_shared_bucket_access": true,
            "server":"http://192.168.33.20:8091",
            "bucket":"data-bucket",
            "username":"data-bucet",
            "password": "passwrd"
        }
    }
}
```

### Log output
```
2017-12-08T01:28:24.926Z Enabling logging: [CRUD+ Cache+ HTTP+ Changes+ Import+]
2017-12-08T01:28:24.926Z ==== Couchbase Sync Gateway/2.0.0(687;a8ad166) ====
2017-12-08T01:28:24.926Z requestedSoftFDLimit >= currentHardFdLimit (90000 >= 65535) capping at 65535
2017-12-08T01:28:24.926Z Configured process to allow 65535 open file descriptors
2017-12-08T01:28:24.926Z Opening db /db as bucket "data-bucket", pool "default", server <http://192.168.33.20:8091>
2017-12-08T01:28:24.926Z GoCBCustomSGTranscoder Opening Couchbase database data-bucket on <http://192.168.33.20:8091> as user "data-bcket"
panic: value method github.com/couchbase/sync_gateway/base.CouchbaseBucketGoCB.CouchbaseServerVersion called using nil *CouchbaseBucketGoCB pointer

goroutine 1 [running]:
github.com/couchbase/sync_gateway/base.(*CouchbaseBucketGoCB).CouchbaseServerVersion(0x0, 0x19, 0xc2e77d, 0x7, 0xc420013cd0, 0xb, 0x0)
        <autogenerated>:1 +0xea
github.com/couchbase/sync_gateway/base.GetBucket(0xc420015660, 0x19, 0xc2e77d, 0x7, 0xc420013cd0, 0xb, 0x0, 0x0, 0x1083c20, 0xc420085e60, ...)
        /home/couchbase/jenkins/workspace/sgw-unix-build@2/2.0.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/base/bucket.go:517 +0x3be
github.com/couchbase/sync_gateway/db.ConnectToBucket.func1(0x0, 0xc420181740, 0xc420181718, 0x46952b, 0x11fde60)
        /home/couchbase/jenkins/workspace/sgw-unix-build@2/2.0.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/db/database.go:145 +0x70
github.com/couchbase/sync_gateway/base.RetryLoop(0xc4200167e0, 0x2a, 0xc42012c3f0, 0xc42010ffa0, 0x1, 0xc4200167e0, 0x2a, 0x17)
        /home/couchbase/jenkins/workspace/sgw-unix-build@2/2.0.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/base/util.go:326 +0x194
github.com/couchbase/sync_gateway/db.ConnectToBucket(0xc420015660, 0x19, 0xc2e77d, 0x7, 0xc420013cd0, 0xb, 0x0, 0x0, 0x1083c20, 0xc420085e60, ...)
        /home/couchbase/jenkins/workspace/sgw-unix-build@2/2.0.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/db/database.go:164 +0x199
github.com/couchbase/sync_gateway/rest.(*ServerContext)._getOrAddDatabaseFromConfig(0xc42018a740, 0xc420085e60, 0xc42018a700, 0x40af70, 0xc420181eb0, 0x0)
        /home/couchbase/jenkins/workspace/sgw-unix-build@2/2.0.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/rest/server_context.go:376 +0x842
github.com/couchbase/sync_gateway/rest.(*ServerContext).getOrAddDatabaseFromConfig(0xc42018a740, 0xc420085e60, 0xc420181e00, 0x0, 0x0, 0x0)
        /home/couchbase/jenkins/workspace/sgw-unix-build@2/2.0.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/rest/server_context.go:256 +0x9c
github.com/couchbase/sync_gateway/rest.(*ServerContext).AddDatabaseFromConfig(0xc42018a740, 0xc420085e60, 0xc420181eb0, 0x0, 0x1)
        /home/couchbase/jenkins/workspace/sgw-unix-build@2/2.0.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/rest/server_context.go:706 +0x3a
github.com/couchbase/sync_gateway/rest.RunServer(0xc420085d40)
        /home/couchbase/jenkins/workspace/sgw-unix-build@2/2.0.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/rest/config.go:770 +0x1ad
github.com/couchbase/sync_gateway/rest.ServerMain(0xc400000000)
        /home/couchbase/jenkins/workspace/sgw-unix-build@2/2.0.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/rest/config.go:828 +0x47
main.main()
        /home/couchbase/jenkins/workspace/sgw-unix-build@2/2.0.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/main.go:40 +0xbe
```

### Expected behavior
A log message should be logged saying that the bucket could not be authenticated  with the provided credentials

### Actual behavior
SG crashed

### Steps to reproduce
1.  Provide invalid username and password for the bucket
2.  Start SG

--- a/base/bucket.go
+++ b/base/bucket.go
@@ -511,9 +511,12 @@ func GetBucket(spec BucketSpec, callback sgbucket.BucketNotifyFn) (bucket Bucket
 			panic(fmt.Sprintf("Unexpected CouchbaseDriver: %v", spec.CouchbaseDriver))
 		}
 
-		if pkgerrors.Cause(err) == gocb.ErrAuthError {
-			Warn("Unable to authenticate: %v", err)
-			return nil, ErrFatalBucketConnection
+		if err != nil {
+			if pkgerrors.Cause(err) == gocb.ErrAuthError {
+				Warn("Unable to authenticate"+suffix+": %v", err)
+				return nil, ErrFatalBucketConnection
+			}
+			return nil, err
 		}
 
 		// If XATTRS are enabled via enable_shared_bucket_access config flag, assert that Couchbase Server is 5.0

--- a/base/bucket.go
+++ b/base/bucket.go
@@ -511,6 +511,11 @@ func GetBucket(spec BucketSpec, callback sgbucket.BucketNotifyFn) (bucket Bucket
 			panic(fmt.Sprintf("Unexpected CouchbaseDriver: %v", spec.CouchbaseDriver))
 		}
 
+		if pkgerrors.Cause(err) == gocb.ErrAuthError {
+			Warn("Unable to authenticate: %v", err)
+			return nil, ErrFatalBucketConnection
+		}
+
 		// If XATTRS are enabled via enable_shared_bucket_access config flag, assert that Couchbase Server is 5.0
 		// or later, otherwise refuse to connect to the bucket since pre 5.0 versions don't support XATTRs
 		if spec.UseXattrs {

--- a/base/bucket_gocb_test.go
+++ b/base/bucket_gocb_test.go
@@ -1944,6 +1944,18 @@ func TestCouchbaseServerVersion(t *testing.T) {
 
 }
 
+func TestCouchbaseServerIncorrectLogin(t *testing.T) {
+
+	if UnitTestUrlIsWalrus() {
+		t.Skip("This test only works against Couchbase Server")
+	}
+
+	// Bad auth creds cause a fatal error with logs indicating the reason why.
+	_, err := GetBucketWithInvalidUsernamePassword(DataBucket)
+	assert.Equals(t, err, ErrFatalBucketConnection)
+
+}
+
 func createTombstonedDoc(bucket *CouchbaseBucketGoCB, key, xattrName string) {
 
 	// Create document with XATTR

--- a/base/util_testing.go
+++ b/base/util_testing.go
@@ -143,7 +143,7 @@ func GetBucketOrPanicCommon(bucketType CouchbaseBucketType) TestBucket {
 
 }
 
-func GetBucketWithInvalidUsernamePassword(bucketType CouchbaseBucketType) (*TestBucket, error) {
+func GetBucketWithInvalidUsernamePassword(bucketType CouchbaseBucketType) (TestBucket, error) {
 
 	spec := GetTestBucketSpec(bucketType)
 
@@ -156,11 +156,8 @@ func GetBucketWithInvalidUsernamePassword(bucketType CouchbaseBucketType) (*Test
 
 	// Attempt to open a test bucket with invalid creds. We should expect an error.
 	bucket, err := GetBucket(spec, nil)
-	if err != nil {
-		return nil, err
-	}
+	return TestBucket{Bucket: bucket}, err
 
-	return &TestBucket{Bucket: bucket}, nil
 }
 
 // Convenience function that will cause a bucket to be created if it doesn't already exist.

--- a/base/util_testing.go
+++ b/base/util_testing.go
@@ -123,9 +123,8 @@ func GetBucketOrPanicCommon(bucketType CouchbaseBucketType) TestBucket {
 			// Create a brand new bucket
 			// TODO: in this case, we should still wait until it's empty, just in case there was somehow residue
 			// TODO: in between deleting and recreating it, if it happened in rapid succession
-			createBucketErr := tbm.CreateTestBucket()
-			if createBucketErr != nil {
-				panic(fmt.Sprintf("Could not create bucket.  Spec: %+v Err: %v", spec, createBucketErr))
+			if err := tbm.CreateTestBucket(); err != nil {
+				panic(fmt.Sprintf("Could not create bucket.  Spec: %+v Err: %v", spec, err))
 			}
 		}
 
@@ -142,6 +141,26 @@ func GetBucketOrPanicCommon(bucketType CouchbaseBucketType) TestBucket {
 
 	return TestBucket{Bucket: bucket}
 
+}
+
+func GetBucketWithInvalidUsernamePassword(bucketType CouchbaseBucketType) (*TestBucket, error) {
+
+	spec := GetTestBucketSpec(bucketType)
+
+	// Override spec's auth with invalid creds
+	spec.Auth = TestAuthenticator{
+		Username:   "invalid_username",
+		Password:   "invalid_password",
+		BucketName: spec.BucketName,
+	}
+
+	// Attempt to open a test bucket with invalid creds. We should expect an error.
+	bucket, err := GetBucket(spec, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &TestBucket{Bucket: bucket}, nil
 }
 
 // Convenience function that will cause a bucket to be created if it doesn't already exist.
@@ -222,8 +241,13 @@ func (tbm *TestBucketManager) OpenTestBucket() (bucketExists bool, err error) {
 	username, password, _ := tbm.BucketSpec.Auth.GetCredentials()
 	bucket, err := tbm.Cluster.OpenBucket(tbm.BucketSpec.BucketName, password)
 	if err != nil {
-		// Let's assume that if there is an error opening the bucket, it's just because the
-		// bucket does not exist
+		// Authentication failure should return an explicit error as we can't continue from here.
+		if pkgerrors.Cause(err) == gocb.ErrAuthError {
+			log.Printf("Unable to authenticate as %s: %v", username, err)
+			return false, err
+		}
+
+		// There could be other errors here, but we assume the bucket doesn't exist, and may be able to continue.
 		// TODO: should check returned error type
 		log.Printf("GoCB error opening bucket: %v", err)
 		return false, nil

